src/internal/pjs/api_server.go

```diff
@@ -111,3 +112,92 @@ func (a *apiServer) InspectJob(ctx context.Context, req *pjs.InspectJobRequest)
 		},
 	}, nil
 }
+
+func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error) {
+	ctx := srv.Context()
+	req, err := srv.Recv()
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	if req.Queue == nil {
+		return status.Errorf(codes.InvalidArgument, "first message must pick Queue")
+	}
+
+	for {
+		var jobID pjsdb.JobID
+		if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+			jid, err := pjsdb.DequeueAndProcess(ctx, sqlTx, req.Queue.Id)
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
+			jobID = jid
+			return nil
+		}); err != nil {
+			// await here
+			if errors.As(err, &pjsdb.DequeueFromEmptyQueueError{}) {
+				time.Sleep(100 * time.Millisecond)
+				continue
+			}
+		}
+		var inputsID []fileset.ID
+		if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+			job, err := pjsdb.GetJob(ctx, sqlTx, jobID)
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
+			inputsID = job.Inputs
+			return nil
+		}); err != nil {
+			return err
+		}
+		var inputs []string
+		for _, filesetID := range inputsID {
+			inputs = append(inputs, filesetID.HexString())
+		}
+		if err := func() error {
+			if err := srv.Send(&pjs.ProcessQueueResponse{
+				Context: "context is not implemented",
+				Input:   inputs,
+			}); err != nil {
+				return errors.EnsureStack(err)
+			}
+			req, err := srv.Recv()
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
+			if req.Result == nil {
+				return status.Errorf(codes.InvalidArgument, "expected Result. HAVE: %v", req)
+			}
+			if req.GetFailed() {
+				if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+					if err := pjsdb.CompleteError(ctx, sqlTx, jobID, pjs.JobErrorCode_FAILED); err != nil {
+						return errors.EnsureStack(err)
+					}
+					return nil
+				}); err != nil {
+					return err
+				}
+			} else if out := req.GetSuccess(); out != nil {
+				if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+					if err := pjsdb.CompleteOk(ctx, sqlTx, jobID, out.Output); err != nil {
+						return errors.EnsureStack(err)
+					}
+					return nil
+				}); err != nil {
+					return err
+				}
+			}
+			return nil
+		}(); err != nil {
+			if errTx := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+				if errDB := pjsdb.CompleteError(ctx, sqlTx, jobID, pjs.JobErrorCode_DISCONNECTED); errDB != nil {
```

Comment:
What's the rational for wrapping the above code in an anonymous function instead of just handling the individual error cases?